### ISSUE TITLE
wildcard support on IO Classification

### ISSUE THREAD

**Issue by tconrado** [Open-CAS/ocf#410]
There is a clear mention in the documentation of how control evection for folders and file extension. 
Is it possible to use wildcards and files too? so, we can effectively create exclusion/inclusions in the best fashion as possible?

**Comment by mmichal10** [comment]
Hi @tconrado 

thank you for your question. Apparently the documentation is incomplete and not all the rules are described. 

If you need wildcard functionality in your rules based on file name you can use combination of two following rules:
- `file_name_prefix`
- `extension`.

For example a rule matching pattern `foo*.txt` looks like following:
`file_name_prefix:foo&extension:txt&done`

**Comment by tconrado** [comment]
Great news @mmichal10 , thanks for making it clear.

is possible to do something like /srv/instances/202*/december.ldb too?

**Comment by mmichal10** [comment]
`directory:/srv/instances/202&file_name_prefix:december&extension` will match /srv/instances/202*/december.ldb

Unfortunately we don't support classification scenario with subfolders excluded.

**Comment by tconrado** [comment]
thanks again for the response; making use of the same example 
will be effective to combine 2 directory (assuming that the regex is not anchored in root) -- like wildcard again?
taking the folders structure as /srv/[payments, invoices, sales],[1990..2020]/[q1,q2,q3,4]/months
```
10 directory:/srv/payments/201&directory:/q4/&done, 9, 1
11 directory:/srv/payments/202&directory:/q4/&done, 9, 1
12 directory:/srv/payments, 22, 0

```
in an attempt to cache all payments from the last decade for Q4, but never cache any other payments?

**Comment by mmichal10** [comment]
I've checked the code responsible for classification and I found out that my previous answer was wrong. The rule I gave You will not match `/srv/instances/202*/december.ldb`. It will match only `/srv/instances/202/december.ldb`. The reason behind that is that we don't compare paths in directory based classifier code. When new `directory` rule is added, given path is resolved into inode and and later the classification is done basing on inodes comparison. If directory under path given in classification rule doesn't exist, no IO will be classified to this ioclass. 

To summarize: to create valid classification rule based on directory, you have to give **path to existing directory**.

**Comment by tconrado** [comment]
thanks for the clarification, we have on last question now. Assuming that the following directories do exist by the moment of rule activation, will the rule above bellow also classify the subfolders on it? Or will it only classify files on the first level of the folder?
```
**10 directory:/srv/instances, 9, 1**

/srv/instances/2019
/srv/instances/2018
/srv/instances/2020
```

**Comment by mmichal10** [comment]
The rule `directory:/srv/instances` will apply to `srv/instances` and all it's subfolders. If you want to disable caching for one of it's subfoloders (i.e 2018), you have to create config with following rules:
```
9,directory:/srv/instances/2018&done, 9, 0
10,directory:/srv/instances&done, 9, 1

```
But still you have to enumerate absolute path to every single directory you want to exclude

**Comment by tconrado** [comment]
interesting, so we can tell it to never evict from cache some specific files by name in the subfolder (assuming that the previous existence of the folders and subfolder)  and define a different eviction priority for the other files in the subfolders as long the subfolders exist before the load of the rules?

```
existing folders:
/srv/instances/2019
/srv/instances/2018
/srv/instances/2020

rules:
10,directory:/srv/instances&file_name_prefix:summary&done,,1
11,directory:/srv/instances/2020&done,9,1
11,directory:/srv/instances,,0
```

this way, we expect to have the summary file always in the cache, and the rest of the files will never get cached, except for the year of 2020. Is this correct? Thanks in advance for all the support.

**Comment by mmichal10** [comment]
> interesting, so we can tell it to never evict from cache some specific files by name in the subfolder 

Yes. This mechanism is also applicable to whole directories and any other classification rules.

> and define a different eviction priority for the other files in the subfolders as long the subfolders exist before the load of the rules?

You can load a rule with a path to a non-existing directory, however the rule will stay inactive (no IO will be classified with this rule) as long as the directory doesn't exist. But you can create it after loading the rule. 

> this way, we expect to have the summary file always in the cache, and the rest of the files will never get cached, except for the year of 2020. Is this correct?

Yes, it is correct

**Comment by tconrado** [comment]
thanks so much for all the responses. It was very useful. best regards!

**Comment by tconrado** [comment]
I'm sorry to re-open this, but I could not find myself the answer and testing became complicated;

I did some reading in linux-open-cas code to better understand it.
seems that you guys use the the function kernel_path to grab the inode of the the folder and build the queue, and probably, for every file that is processed in the IO classification, it check against the parent inode info and if it does match any in the queue, it will follow that specific rule;

my question is: if we build a folder, full of symlinks (softlinks) to folders that we want to have cached, would it work?

**Comment by mmichal10** [comment]
Unfortunately it will not. You can create rule with paths to each single directory you want to cache, or with path to one common parent directory. But either way, it has to be a absolute path to the actual directory you want to be cached, not to the symlink.
